### ISSUE TITLE
Add `impersonated_id` to `HELLO` to provide user impersonation

### ISSUE THREAD

**Issue by konsultaner** [wamp-proto/wamp-proto#297]
@ecorm started to define WAMP-SCRAM in #135. It includes the possibility to impersonate users by setting an authorization identity in conjunction with the `authid` in the `HELLO` message. The WAMP-SCRAM draft describes user impersonation as follows:

>Impersonation allows a priviledged user to log in using his/her credentials, but act as a different identity within the application, using the permissions of that different identify. A >use case for this feature would be an administrator authenticating using his/her credentials and then customizing a user's settings as if the administrator were that user. Another >use case would be a technical support representative remotely logging into the server, but acting as if he/she is the customer.
>
>SCRAM supports the SASL authorization identity defined in [RFC4422, section 3.4.1](https://tools.ietf.org/html/rfc4422#section-3.4.1). This SASL authorization identify allows a client to impersonate a user.
>
>How the application behaves for an impersonated user, or whether user impersonation is even permitted, is a local matter outside the scope of this document.

It would be nice to see this feature work with other authentication methods as well. The `HELLO.Details` would be extended by an `impersonated_id` field as explained in the WAMP-SCRAM draft:

>```
>[1, "realm1",
>    {
>        "roles": ...,
>        "authmethods": ["wamp-scram-pbkdf2"],
>        "authid": "user",
>        "nonce": "rOprNGfwEbeRWgbNEkqO",
>        "impersonated_id": null,
>        "gs2_cbind_flag": null,
>    }
>]
>```

In my oppinion in the following scenarios

- The user associated with authid does not have permission to impersonate the given impersonated_id.
- Ther server recognizes authid but does not recognize impersonated_id.
- The server does not recognize the given authid.

the router should send a mock `Challenge` and fail the authentication process after receiving the `AUTHENTICATE` message.

**Comment by oberstet** [comment]
I am -1 on this.

**Comment by konsultaner** [comment]
@ecorm @meejah @KSDaemon @Paranaix , Because this came up in the WAMP-SCRAM issue (#135)
If you could +1 or -1 this issue, would be nice. Because I need to implement it soon and I don't know wheather this has a chance to be part of the specs or will be a custom implementation.

**Comment by KSDaemon** [comment]
Well... First of all, we should be consistent with naming. If we mostly use merged writing, than we should avoind messing it up with snake cases. I mean `authmethods` vs `impersonated_id`. This should be `impersonatedid`.

Regarding issue itself... I don't like this at 100%. Something is worrying me. May be we can extend `HELLO.Details` dictionary with something like `extra` key, that can hold such additional info?...

So, basically, i think we need to examine and discuss this more... -1 in this case right now...

**Comment by ecorm** [comment]
@konsultaner  I'm neutral on this, as we don't need this feature in our application.

If you can't wait, perhaps you could somehow implement this at the application level?

**Comment by oberstet** [comment]
closing: not enough interest - and quite complex / tricky security implications
